x/posts/keeper/handler.go

```diff
@@ -36,16 +38,27 @@ func NewHandler(keeper Keeper) sdk.Handler {
 	}
 }
 
+// ComputeID returns a sha256 hash of the msg's json representation
+// nolint: interfacer
+func ComputeID(msg types.MsgCreatePost) types.PostID {
+	bz, err := msg.MarshalJSON()
+	if err != nil {
+		panic(err)
+	}
+	hash := sha256.Sum256(bz)
+	return types.PostID(hex.EncodeToString(hash[:]))
+}
```

Comment:
While this implementation surely works, it has a major flaw in my opinion: clients will not be able to use it properly.

The advantage of having a deterministic id is that even client application will be able to generate them before event sending them to the chain. If we require them to know the concept and structure of a `MsgCreatePost`, this might make such pre-generation process harder for them. 

For this reason, I think that we should define a better way of generating the id. What I was thinking we might want to do is the following: 

1. Remove the `ID` field from the `NewPost` method. 
2. Inside the `NewPost` method body, create the proper ID hashing all the other data together (`ParentId`, `Message`, `AllowsComments`, `Subspace`, `OptionalData`, `CreationTime` and `Creator`). 

This way, the ID will be generated when creating the post object in a way that it's easily reproducible by dApps.

Reply:
I agree. With so many possible clients the last thing to do is surely complicate their lives.

Reply:
Actually, since the `NewPost` method has 85 reference around the code, I've preferred to *not* edit it but edit only the `ComputeID` method instead.

Reply:
@bragaz Ok, that's a nice choice. I would still make so that `ComputeID` does not need a `MsgCreatePost` tho, but instead the whole set of data that should be used. This is to avoid the problem described above. 

Currently if we use the implementation you have written, a client library would hardly use it. Right now it should behave as follows: 

1. Create a JSON `MsgCreatePost`
2. Create a `Post` object

So, the client would have to know the format of both the `MsgCreatePost` and `Post` objects, which is not that good. Instead, what we should do is the following: 

1. Create a `Post` object. 
2. Hash all of its content in the JSON format, excluding the ID. 

So, for example: 

```javascript
const post = {
  "parent_id": "",
  "message": "Desmos is great!",
  "allows_comments": true,
  "subspace": "4e188d9c17150037d5199bbdb91ae1eb2a78a15aca04cb35530cccb81494b36e",
  "creator": "desmos1w3fe8zq5jrxd4nz49hllg75sw7m24qyc7tnaax",
  "creation_date": "2020-01-01T12:00:00Z"
}

const postID = sha256(jsonEncode(post))
```

This can probably be achieved using the following code: 

```golang
func NewPost(id, parentID PostID, message string, allowsComments bool, subspace string,
  optionalData map[string]string, created time.Time, creator sdk.AccAddress) Post {
  post := Post{ ... }
  
  postBz, err := json.Marshal(&post)
  if err != nil {
    panic(err)
  }
  
  postID := sha256.Sum256(postBz)
  post.PostID = PostID(hex.EncodeToString(hash[:]))
  return post
}
```

This would make sure that even client libraries can reproduce these steps without knowing the format of a `MsgCreatePost`. Also, thanks to it, we would be able to get a post id based on its content instead of knowing the original `MsgCreatePost` which is not available inside some places like the genesis.

Reply:
We should move this to another PR @RiccardoM